kvclient: cache the sorted list of replicas

andrewbaptist opened the pull request:
Calling OptimizeReplicaOrder is expensive and can show up as a hotspot in many high QPS workloads. This PR caches the sorted list on a per-range basis and updates the cache at most once every 3s.

Epic: none

Release note: None

Commit:
kvclient: cache the sorted list of replicas

Calling OptimizeReplicaOrder is expensive and can show up as a hotspot
in many high QPS workloads. This PR caches the sorted list on a
per-range basis and updates the cache at most once every 3s.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114033)
<!-- Reviewable:end -->

andrewbaptist commented:
@nvanbenschoten FYI - still a work in progress, but I think the change here will be pretty straighforward to get working well.

Commit:
kvclient: replace ReplicaSlice with ReplicaSet for routing

This PR replaces ReplicaSlice with ReplicaSet outside of the
DistSender. ReplicaSlice is an internal implementation which
contains information only required for sorting the replicas.
Outside of DistSender the additional sorting information is
unnecessary and unused.

Epic: none
Informs: #112351

Release note: None

andrewbaptist commented:
Running against one node (after writing ~2 minutes)
```
./cockroach workload run kv --read-percent=100 --follower-read-percent=100 --write-seq=R394703 --concurrency 200 'postgres://root@10.142.1.135:26257?sslmode=disable'
```

Small but measurable performance. With the additional changes to include the health, it removes a call to NodeDialer.ConnHealth (0.7%) and therefore should be an even larger improvement AND have better behavior.

Calls to DistSender.sendToReplicas:
New:
![image](https://github.com/cockroachdb/cockroach/assets/46975918/247f7f3f-db5d-4a67-9b2e-c1dca42bbc1f)

Old:
<img width="619" alt="image" src="https://github.com/cockroachdb/cockroach/assets/46975918/ec7cfa6f-a9bd-4ee9-a094-53adef5bc065">

Calls to OptimizeReplicaOrder:
New:
<img width="563" alt="image" src="https://github.com/cockroachdb/cockroach/assets/46975918/61923999-d517-400e-a0a5-e85a4cf0e272">

Old:
<img width="548" alt="image" src="https://github.com/cockroachdb/cockroach/assets/46975918/a6d47e97-7516-4594-b9c4-ab24e091339f">

andrewbaptist commented:
@nvanbenschoten or @arulajmani if you get a chance to review this (or comment on just the first commit which is in #114429) I can try and get this merged this week which will unblock adding checks as the default setting.

Commit:
kvclient: remove EvictByKey from the interface

Remove the EvictByKey from the interface. This replaces everything in
the cache, and it is better to update entries through an EvictionToken.

Epic: none

Release note: None

Commit:
kvclient: change GetCached to return RangeInfo

The internal CacheEntry inside the RangeCache should not be publicly
exposed. This commit changes GetCached to return the public RangeInfo
instead. This is a test only menthod.

Epic: none

Release note: None

Commit:
kvclient: remove unnecessary getters

Now that the RangeCache.CacheEntry is private, we no longer need all the
getters for the different fields. This simplifies some test code that
was converting back and froth from pointers, structs.

Epic: none

Release note: None

Commit:
kvclient: don't return a cache.Entry from GetCachedOverlapping

The cache.Entry is not intended to be public, instead return a
RangeInfo.

Epic: none

Release note: None

Commit:
kvclient: remove MakeEvictionToken from the interface

This method was only used internally and is simple to create manually
now. It was confusing to have it on the public interface as its not the
right way to create a token.

Epic: none

Release note: None

Commit:
kvclient: make cacheEntry private

CacheEntry is not intended to be exposed externally. This commit makes
it a private type in the RangeCache.

Epic: none

Release note: None

Commit:
kvclient: pull out the method to sort replicas

In preperation for caching the sorted replicas, pull out the function to
generate the two sorted lists. This is meant to be an intermediate
commit as it will have worse performance characteristics since it always
computes the list twice.

Epic: none

Release note: None
